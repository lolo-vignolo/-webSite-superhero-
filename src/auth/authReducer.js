import React from 'react'
import { types } from '../types/types';

//estado si el usuario esta autenticado
// const state = {
//     name: "Lorenzo",
//     logged: true
// }


// const ejAction = {

//     type : "login",
//     payload : {
//         name : "newname",
//         logged : true or false
//     }
// }


export const authReducer = (state = {}, action ) => {
   

    switch (action.type) {
        case  types.login:
            
            return {
              ...action.payload,
              logged : true
          }
          
        case types.logout:
            
            return{
                logged:false
            }

        default:
        return state;



    }
}

