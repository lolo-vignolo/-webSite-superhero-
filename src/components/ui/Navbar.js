import React, { useContext } from 'react';
import { Link, NavLink , useHistory } from 'react-router-dom';
import { AuthContext } from '../../auth/AuthContext';
import { types } from '../../types/types';

export const Navbar = () => {

    const {user: {name}, dispatch} =  useContext(AuthContext);

    // no puedo usar history como en el login por que el logout este dentro de las
    //dashboardroutes las cuales estan creadas dentro de un Navar y no dentro de un 
    // router el cual podría brindarnos ese history, por eso react creó el hook useHistory
    // para estas ocaciones, donde esta el replece, y otros usos.

    const history = useHistory();

    const handleLogout = ()=>{

        const logout = {
            type : types.logout,
        }
        dispatch(logout)
        history.replace("/login")
        

    }



    return (
        <nav className="navbar navbar-expand-sm navbar-dark bg-dark ">
          <div className = "navbar-nav p-2">
            <Link 
                className="navbar-brand " 
                to="/"
            >
                Asociaciones
            </Link>
            </div>

            <div className="navbar-collapse">
                <div className="navbar-nav">

                    <NavLink 
                        activeClassName="active"
                        className="nav-item nav-link" 
                        exact
                        to="/marvel"
                    >
                        Marvel
                    </NavLink>

                    <NavLink 
                        activeClassName="active"
                        className="nav-item nav-link" 
                        exact
                        to="/dc"
                    >
                        DC
                    </NavLink>

                    <NavLink 
                        activeClassName="active"
                        className="nav-item nav-link" 
                        exact
                        to="/search"
                    >
                        Search
                    </NavLink>
                </div>
            </div>

            {/* <div className="navbar-collapse collapse w-100 order-3 dual-collapse2"> */}
                <ul className="navbar-nav ml-auto">

                    <span className = "nav-item nav-link text-info" >
                            {name}
                    </span>
                
                    <button 
                    
                        className="nav-item nav-link btn" 
                        onClick = {handleLogout}
                    >
                        Logout
                    </button>
                </ul>
            {/* </div> */}
          
        </nav>
    )
}