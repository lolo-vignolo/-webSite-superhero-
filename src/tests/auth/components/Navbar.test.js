import { mount } from "enzyme"
import { MemoryRouter, Router } from "react-router"
import { AuthContext } from "../../../auth/AuthContext"
import { Navbar } from "../../../components/ui/Navbar"
import { types } from "../../../types/types"

describe('pruebas en  componente <Navbar />', () => {

    const historyMock = {
        push:jest.fn(),
        replace : jest.fn(),
        location: {},
        lsiten:jest.fn(),
        createHref: jest.fn(),
    }

    test('should show it properly ', () => {

        const contextValue = {
            dispatch: jest.fn(),
            user:{
                logged: true,
                name : "Lorenzo"
            }
        }

        const wrapper = mount ( 
            <AuthContext.Provider value = {contextValue} >
            <MemoryRouter> 
                <Navbar />
            </MemoryRouter>
        </AuthContext.Provider>
        
         )
        

         expect(wrapper).toMatchSnapshot();
         expect(wrapper.find(".text-info").text().trim()).toBe("Lorenzo")
    });

    test('should calls to logout and history ', () => {

        const contextValue = {
            dispatch: jest.fn(),
            user:{
                logged: true,
                name : "Lorenzo"
            }
        }

        const wrapper = mount ( 
            <AuthContext.Provider value = {contextValue} >
            <MemoryRouter> 
                <Router history = {historyMock}>
                    <Navbar />
                </Router>
            </MemoryRouter>
        </AuthContext.Provider>
        
         );

         afterEach(()=>{
            jest.clearAllMocks();
        });


         //find encuentra al button, que dentro tiene un prop onClick, que al tocar este llama a una funcion ().
        wrapper.find("button").prop("onClick")();

        expect (contextValue.dispatch).toHaveBeenCalledWith({
            type:types.logout
        });

        expect(historyMock.replace).toHaveBeenCalledWith("/login");



       


        
    })
    
    
})
