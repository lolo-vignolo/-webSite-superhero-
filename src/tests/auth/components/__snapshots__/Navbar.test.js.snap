// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en  componente <Navbar /> should show it properly  1`] = `
<nav
  className="navbar navbar-expand-sm navbar-dark bg-dark "
>
  <div
    className="navbar-nav p-2"
  >
    <a
      className="navbar-brand "
      href="/"
      onClick={[Function]}
    >
      Asociaciones
    </a>
  </div>
  <div
    className="navbar-collapse"
  >
    <div
      className="navbar-nav"
    >
      <a
        aria-current={null}
        className="nav-item nav-link"
        href="/marvel"
        onClick={[Function]}
      >
        Marvel
      </a>
      <a
        aria-current={null}
        className="nav-item nav-link"
        href="/dc"
        onClick={[Function]}
      >
        DC
      </a>
      <a
        aria-current={null}
        className="nav-item nav-link"
        href="/search"
        onClick={[Function]}
      >
        Search
      </a>
    </div>
  </div>
  <ul
    className="navbar-nav ml-auto"
  >
    <span
      className="nav-item nav-link text-info"
    >
      Lorenzo
    </span>
    <button
      className="nav-item nav-link btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </ul>
</nav>
`;
